log.js

```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env node
+
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var console = require('console');
+var util = require('util');
+module.exports = Logger;
+
+function Logger(options) {
+    if (!(this instanceof Logger)) {
+        return new Logger(options);
+    }
+
+    var self = this;
+    self.options = options;
```

Comment:
Alright, next pattern: Don’t retain an options object on an instance. Extract the options and drop the options object. In this case, keep a `json` property on the logger.

V8 uses hidden classes. If the parameters of a method always have the same pattern of hidden classes, they can be JIT optimized, and if the JIT code encounters a different pattern, the method gets deoptimized. Options objects tend to not come from the same construction sites and have different hidden classes. By extracting the properties here, we ensure that the Logger always produces the same hidden class and the display function can be optimized.

Of course this doesn’t matter for TCurl, and this is :ok: to land, but keep it in mind in TChannel.